src/sysupdate/sysupdate-resource.c

```diff
@@ -546,7 +550,19 @@ int resource_resolve_path(
                         return log_error_errno(SYNTHETIC_ERRNO(EPERM),
                                                "Block device is not allowed when using --root= mode.");
 
-                r = get_block_device_harder("/usr/", &d);
+                r = readlink_malloc("/run/systemd/volatile-root", &orig_root);
+                if (r == -ENOENT) { /* volatile-root not found */
+                        r = get_block_device_harder("/usr/", &d);
+                } else if (r < 0)
+                        return log_error_errno(r, "Failed to read symlink /run/systemd/volatile-root: %m");
+                else {
+                        mode_t m;
+                        r = device_path_parse_major_minor(orig_root, &m, &d);
```

Comment:
hmm, so gpt-auto-generator uses device_path_parse_major_minor() because it runs before udev does, and thus cannot rely that the `/dev/block/x:y` symlinks are in place yet. But sysupdate doesn't run that early, hence it should not do this dance, and instead just use the path as is. i.e. use a simple `stat(orig_roog, &st)` here, that follows the symlink. And then validate that it indeed checks out to be a block device.

or in other words: I think the fact that /run/systemd/volatile-root only contains a major/minor based symlink path is not something we should rely on except where we really really have to. Whenever we can, we should just assume it's a path to a device node (or a symlink to it)

Reply:
Thanks for the feedback -- this was very helpful. I've pushed an update that aims to address your concerns.